src/Tasks/GenerateLauncher.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Reflection;
+using System.Runtime.Versioning;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Shared;
+using Microsoft.Build.Tasks.Deployment.Bootstrapper;
+using Microsoft.Build.Tasks.Deployment.ManifestUtilities;
+using Microsoft.Build.Utilities;
+
+namespace Microsoft.Build.Tasks
+{
+    /// <summary>
+    /// Generates a bootstrapper for ClickOnce deployment projects.
+    /// </summary>
+    public sealed class GenerateLauncher : TaskExtension
+    {
+        private const string LAUNCHER_EXE = "Launcher.exe";
+        private const string ENGINE_PATH = "Engine"; // relative to ClickOnce bootstrapper path
+
+        #region Properties
+
+        public ITaskItem EntryPoint { get; set; }
+
+        public string LauncherPath { get; set; }
+
+        public string OutputPath { get; set; }
+
+        public string VisualStudioVersion { get; set; }
+
+        [Output]
+        public ITaskItem OutputEntryPoint { get; set; }
+
+        [Output]
+        public string FrameworkName { get; set; }
+
+        [Output]
+        public string FrameworkVersion { get; set; }
+        #endregion
+
+        public override bool Execute()
+        {
+            if (LauncherPath == null)
+            {
+                // Launcher lives next to ClickOnce bootstrapper.
+                // GetDefaultPath obtains the root ClickOnce boostrapper path.
+                LauncherPath = Path.Combine(
+                    Microsoft.Build.Tasks.Deployment.Bootstrapper.Util.GetDefaultPath(VisualStudioVersion),
+                    ENGINE_PATH,
+                    LAUNCHER_EXE);
+            }
+
+            if (EntryPoint == null)
+            {
+                return false;
+            }
+
+            // Get Framework name and version.
+            // Launcher-based manifest generation has to use Framework elements that match Launcher identity.
+            Assembly a = Assembly.UnsafeLoadFrom(LauncherPath);
```

Comment:
Won't this lock the launcher assembly for the lifetime of the MSBuild process? That can be longer than the build, with node reuse.

Can you use `System.Reflection.Metadata` for this, to treat the file as just a bag of bytes?

Reply:
Yes, I will update this.

Reply:
@rainersigwald this seems non-trivial. The code runs under .NET FX, which does not have System.Reflection.Metadata.MetadataReader class. I could try to use IMetadataImport2 interface and extend the existing MetadataReader code: https://github.com/dotnet/msbuild/blob/037c1214b5f49b370d19e2a6657a01aedc16ef26/src/Tasks/ManifestUtil/MetadataReader.cs#L293

However, the only reason for getting Framework name and version from Launcher.exe is to ensure that generated manifest has the same FX target version. Launcher targets 4.5 and this will never change. Is it worth extending MetadataReader class with something that is not immediately needed? The alternative is to hard-code this value in task or target, which isn't really appealing.

Reply:
I've pushed a fix for this issue.